crates/ln-dlc-node/src/networking/tungstenite.rs

```diff
@@ -0,0 +1,169 @@
+use std::hash::Hasher;
+use std::ops::ControlFlow;
+use std::ops::Deref;
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering;
+use tokio::sync::mpsc;
+use tokio::sync::mpsc::UnboundedReceiver;
+use tokio_tungstenite_wasm::Message;
+use tokio_tungstenite_wasm::WebSocketStream;
+use tracing::error;
+
+static ID_COUNTER: AtomicU64 = AtomicU64::new(0);
+
+pub async fn connect_outbound<PM>(
+    peer_manager: PM,
+    node_info: NodeInfo,
+) -> Option<impl Future<Output = ()>>
+where
+    PM: Deref + 'static + Send + Sync + Clone,
+    PM::Target: APeerManager<Descriptor = DynamicSocketDescriptor>,
+{
+    let url = &format!(
+        "ws://{}:{}",
+        node_info.address.ip(),
+        node_info.address.port()
+    );
+    let mut ws = tokio_tungstenite_wasm::connect(url)
+        .await
+        .expect("Can't connect");
+
+    let (task_tx, mut task_rx) = mpsc::unbounded_channel();
+    let mut descriptor = DynamicSocketDescriptor::Tungstenite(SocketDescriptor {
+        tx: task_tx,
+        id: ID_COUNTER.fetch_add(1, Ordering::AcqRel),
+    });
+
+    if let Ok(initial_send) = peer_manager.as_ref().new_outbound_connection(
+        node_info.pubkey,
+        descriptor.clone(),
+        Some(node_info.address.into()),
+    ) {
+        ws.send(Message::Binary(initial_send))
+            .await
+            .expect("failed to send initial");
+
+        Some(async move {
+            let mut emit_read_events = true;
+            loop {
+                match process_messages(
+                    &peer_manager,
+                    &mut task_rx,
+                    &mut ws,
+                    &mut descriptor,
+                    &mut emit_read_events,
+                )
+                .await
+                {
+                    Ok(ControlFlow::Break(())) => break,
+                    Ok(ControlFlow::Continue(())) => (),
+                    Err(err) => {
+                        error!("Disconnecting websocket with error: {err}");
+                        let _ = ws.close().await;
+                        peer_manager.as_ref().socket_disconnected(&descriptor);
+                        peer_manager.as_ref().process_events();
+                        break;
+                    }
+                }
+            }
+        })
+    } else {
+        None
+    }
+}
+
+async fn process_messages<PM>(
+    peer_manager: &PM,
+    task_rx: &mut UnboundedReceiver<BgTaskMessage>,
+    ws: &mut WebSocketStream,
+    descriptor: &mut DynamicSocketDescriptor,
+    emit_read_events: &mut bool,
+) -> Result<ControlFlow<()>, anyhow::Error>
+where
+    PM: Deref + 'static + Send + Sync + Clone,
+    PM::Target: APeerManager<Descriptor = DynamicSocketDescriptor>,
+{
+    let ws_next = if *emit_read_events {
+        Either::Left(ws.next())
+    } else {
+        Either::Right(future::pending())
+    };
+
+    tokio::select! {
+        task_msg = task_rx.recv() => match task_msg.context("rust-lighting SocketDescriptor dropped")? {
```

Comment:
Classic Lightning typo 😛 

```suggestion
        task_msg = task_rx.recv() => match task_msg.context("rust-lightning SocketDescriptor dropped")? {
```

Reply:
Oops :sweat_smile: